net/dns/direct.go

```diff
@@ -324,3 +324,55 @@ func (m directManager) Close() error {
 
 	return nil
 }
+
+func atomicWriteFile(fs pinholeFS, filename string, data []byte, perm os.FileMode) error {
+	var randBytes [12]byte
+	if _, err := rand.Read(randBytes[:]); err != nil {
+		return fmt.Errorf("atomicWriteFile: %w", err)
+	}
+
+	tmpName := fmt.Sprintf("%s.%x.tmp", filename, randBytes[:])
+	defer fs.Remove(tmpName)
+
+	if err := fs.WriteFile(tmpName, data, perm); err != nil {
+		return fmt.Errorf("atomicWriteFile: %w", err)
+	}
+	return fs.Rename(tmpName, filename)
+}
+
+// pinholeFS is a high-level file system abstraction designed just for use
```

Comment:
I don't get the name. What's pinhole?

Reply:
I couldn't think of a name for this. It's an FS, but not a standard open/read/write/close Unix API. It felt like looking through a pinhole. Any better names?

Reply:
Took another attempt at the name, wholeFileFS, as you can only work with whole files.